### PR TITLE
Filter sessions based on new time range.

### PR DESCRIPTION
**Pull request by AkkarinForest**
Some more info in the commits.

### CODE REVIEW

File: app/models/fixed_session.rb [hunk 1]
```diff
@@ -43,6 +43,7 @@ def after_measurements_created
   end
 
   def update_end_time!
+    # Measurement.time is a local time, so this are both local end times:
     self.end_time = self.measurements.maximum('time')
     self.end_time_local = self.measurements.maximum('time')
     self.last_measurement_at = DateTime.current
```
Comment (AkkarinForest) [inline app/models/fixed_session.rb]: We should change that if we add the utc column to measurement. Or just delete `start_time` and `end_time` from Session.

File: app/controllers/api/regions_controller.rb [hunk 2]
```diff
@@ -34,14 +32,8 @@ def show
...
-      data[:time_from] = data[:time_from] || 0
-      data[:time_to] = data[:time_to] || 2359
-
-      data[:day_from] = data[:day_from] || 0
-      data[:day_to] = data[:day_to] || 365
-
-      data[:year_from] = data[:year_from] || 2010
-      data[:year_to] = data[:year_to] || 2050
+      data[:time_from] = data[:time_from] || Time.new(2010).to_i
+      data[:time_to] = data[:time_to] || Time.new(2100).end_of_year.to_i
```
Comment (AkkarinForest) [inline app/controllers/api/regions_controller.rb]: I'm not sure why the defaults here are necessary, I added them just in case, but maybe we don't need them at all.

File: app/models/fixed_session.rb (hunk 1 above)
Comment (3v0k4) [reply app/models/fixed_session.rb]: Could you elaborate more on this please?

File: app/models/session.rb [hunk 3]
```diff
@@ -125,10 +115,28 @@ def self.from_location(data)
...
+      (:time_from BETWEEN start_time_local AND end_time_local)",
+      :time_from => time_from, :time_to => time_to)
+      .local_minutes_range(minutes_of_day(time_from), minutes_of_day(time_to))
+  end
+
   def self.whole_day?(time_from, time_to)
     time_from == FIRST_MINUTE_OF_DAY && time_to == LAST_MINUTE_OF_DAY
   end
 
+  def self.minutes_of_day(time)
```
Comment (3v0k4) [inline app/models/session.rb]: Have you considered having this in another module? It seems like it's not related to Session

File: spec/models/session_spec.rb [hunk 4]
```diff
@@ -1,52 +1,97 @@
...
+
+    it "returns sessions that started before the range and ended after the range" do
+      session = create_session(start_time: Time.new(2010, 1, 1), end_time: Time.new(2010, 1, 4))
+
+      actual = Session.filter_by_time_range(Session.all, Time.new(2010, 1, 2).to_i, Time.new(2010, 1, 3).to_i)
+
+      expect(actual.all).to eq([session])
+    end
+
+    it "does not return sessions that dont everlap with the range" do
```
Comment (3v0k4) [inline spec/models/session_spec.rb]: everlap

File: spec/models/session_spec.rb [hunk 5]
```diff
@@ -1,52 +1,97 @@
...
+
+    it "returns sessions that started before the range and ended after the minutes range" do
+      session = create_session(start_time: Time.new(2010, 1, 2, 1, 1), end_time: Time.new(2010, 1, 2, 1, 4))
+
+      actual = Session.filter_by_time_range(Session.all, Time.new(2010, 1, 1, 1, 2).to_i, Time.new(2010, 1, 3, 1, 3).to_i)
+
+      expect(actual.all).to eq([session])
+    end
+
+    it "does not return sessions that dont everlap with the minutes range even when they are in the date range" do
```
Comment (3v0k4) [inline spec/models/session_spec.rb]: everlap

File: app/models/session.rb [hunk 6]
```diff
@@ -49,26 +47,25 @@ class Session < ActiveRecord::Base
...
-
-    where "
-      (#{field_in_minutes.call('start_time_local')} BETWEEN :start_minutes AND :end_minutes)
-      OR
-      (#{field_in_minutes.call('end_time_local')} BETWEEN :start_minutes AND :end_minutes)
-      OR
-      (:start_minutes BETWEEN #{field_in_minutes.call('start_time_local')} AND #{field_in_minutes.call('end_time_local')})
-    ", :start_minutes => start_minutes, :end_minutes => end_minutes
+  scope :local_minutes_range, lambda { |minutes_from, minutes_to|
+    if !whole_day?(minutes_from, minutes_to)
```
Comment (3v0k4) [inline app/models/session.rb]: Have you considered `unless`?

Comment (3v0k4) [reply app/models/session.rb]: `whole_day?` is another thing that doesn't sound related to Session

File: app/models/session.rb [hunk 7]
```diff
@@ -49,26 +47,25 @@ class Session < ActiveRecord::Base
...
+      field_in_minutes = lambda { |field|
```
Comment (3v0k4) [inline app/models/session.rb]: Wow, we need to come back to this query at some point

File: app/models/session.rb [hunk 8]
```diff
@@ -125,10 +115,28 @@ def self.from_location(data)
...
+    time_to = Time.strptime(time_to.to_s, '%s')
+
+    sessions.where(
+      "(start_time_local BETWEEN :time_from AND :time_to)
+      OR
+      (end_time_local BETWEEN :time_from AND :time_to)
+      OR
+      (:time_from BETWEEN start_time_local AND end_time_local)",
+      :time_from => time_from, :time_to => time_to)
+      .local_minutes_range(minutes_of_day(time_from), minutes_of_day(time_to))
```
Comment (3v0k4) [inline app/models/session.rb]: Would it be possible to have one query instead of two? I mean merging `filter_by_time_range` and `local_minutes_range` together?

File: app/models/session.rb [hunk 9]
```diff
@@ -125,10 +115,28 @@ def self.from_location(data)
     where(:streams => {:id => streams_ids.uniq})
   end
 
+  def self.filter_by_time_range(sessions, time_from, time_to)
+    time_from = Time.strptime(time_from.to_s, '%s')
+    time_to = Time.strptime(time_to.to_s, '%s')
```
Comment (3v0k4) [inline app/models/session.rb]: We should not allow url params to get this deep into the code unchecked. The controller should take all the params, validate them and convert to what deeper objects need. We should start moving in that direction asap.

This is an outdated example of that (https://github.com/3v0k4/rails_sane_architecture/blob/master/app/forms/comment_form.rb). The readme explains the structure https://github.com/3v0k4/rails_sane_architecture. I'd prolly use https://dry-rb.org/gems/dry-validation/ and https://dry-rb.org/gems/dry-struct/ nowadays for that

File: app/models/measurement.rb [hunk 10]
```diff
@@ -73,9 +70,23 @@ class Measurement < ActiveRecord::Base
...
-      year_range(Date.new(data[:year_from].to_i),Date.new(data[:year_to].to_i + 1) - 1)
+    time_from = Time.strptime(data[:time_from].to_s, '%s')
+    time_to = Time.strptime(data[:time_to].to_s, '%s')
+
+    where("DATE(time) >= ?", time_from.beginning_of_day)
+      .where("DATE(time) <= ?",  time_to.end_of_day)
+      .minutes_range(Session.minutes_of_day(time_from), Session.minutes_of_day(time_to))
+  end)
+
+  scope(:minutes_range, lambda do |minutes_from, minutes_to|
```
Comment (3v0k4) [inline app/models/measurement.rb]: Do we need to keep this query separated from `with_time` or could they be one scope?

File: app/models/fixed_session.rb (hunk 1 above)
Comment (AkkarinForest) [reply app/models/fixed_session.rb]: Sure. So we do:
```
self.end_time = self.measurements.maximum('time')
self.end_time_local = self.measurements.maximum('time')
```

And this `maximum('time')` comes from `Measurement.time` column, which as we know is saved as local time. So the `Session.end_time` is the same as `Session.end_local_time` here. 
I think there is no reason to have this data duplicated in the database. So we should:
- either remove the `end_time` column from the db
- or change `end_time` to store the time in utc. (eg by adding the `utc_time` column to Measurements and using it to create `self.end_time = self.measurements.maximum('utc_time')`)

And I think there is a similar problem might be with `start_time` and `start_local_time`.

File: app/models/session.rb (hunk 8 above)
Comment (AkkarinForest) [reply app/models/session.rb]: I'm not sure how to make it just one query since the `local_minutes_range` is only performed `unless whole_day?`.

But can I ask why you suggested that? Is it about readability? Or about performance? Cause in the latter case I think it doesn't change anything, rails take care of that. They don't execute every sql line immediately. They collect all the code that looks like sql, put it together and only execute when the resulting object is actually needed, right?

File: app/models/fixed_session.rb (hunk 1 above)
Comment (3v0k4) [reply app/models/fixed_session.rb]: Would you mind creating a ticket somewhere and documenting this please?

File: app/models/session.rb (hunk 8 above)
Comment (3v0k4) [reply app/models/session.rb]: Yeah, the query builder is lazy that's correct.

I've stared at the code for some time and I couldn't really get why there were two queries when the column is one. Thus readability is something we could work on.

The perf is something we can take a look later on I guess. But in case we wanted to use indexes here, at the moment the queries are working on
- `end_time_local`
- `start_time_local`
- `EXTRACT(HOUR FROM end_time_local)`
- `EXTRACT(HOUR FROM start_time_local)`
When a SQL function is applied to a column then the index can't be used. In other words, an index on `end_time_local` would not be used for `EXTRACT(HOUR FROM end_time_local)`
